### PR TITLE
[iOS] Returning operations from sync actions

### PR DESCRIPTION
**Pull request by damienpontifex**
Closes #703 by bringing these changes into dev branch first.

Returns NSOperation types from sync operations to allow the use to use them to react to changes. This could be to cancel, or to add these as dependencies to a trailing operation.

### CODE REVIEW

**Comment by azurecla** [comment]
Hi __@damienpontifex__, I'm your friendly neighborhood Azure Pull Request Bot (You can call me AZPRBOT). Thanks for your contribution!
    <span>You've already signed the contribution license agreement. Thanks!</span>
        <p>The agreement was validated by Microsoft and real humans are currently evaluating your PR.</p>

TTYL, AZPRBOT;

File: sdk/iOS/src/MSSyncContext.m [hunk 1]
```diff
@@ -404,54 +406,53 @@ - (void) pullWithQuery:(MSQuery *)query queryId:(NSString *)queryId completion:(
...
+	MSQueuePullOperation *pull = [[MSQueuePullOperation alloc] initWithSyncContext:self
+																			 query:query
+																		   queryId:queryId
+																		maxRecords:maxRecords
+																	 dispatchQueue:writeOperationQueue
+																	 callbackQueue:self.callbackQueue
+																		completion:completion];
+	
+	// Before we can pull from the remote, we need to make sure out table doesn't having pending operations
+	NSArray *tableOps = [self.operationQueue getOperationsForTable:query.table.name item:nil];
```
Comment (phvannor) [inline sdk/iOS/src/MSSyncContext.m]: Keeping this check on the original queue here will incur a performance penalty in the user's app, as now they are waiting for a db read.  What's the rationale for keeping it here?

File: sdk/iOS/src/MSSyncContext.m [hunk 2]
```diff
@@ -404,54 +406,53 @@ - (void) pullWithQuery:(MSQuery *)query queryId:(NSString *)queryId completion:(
...
+			// Long term we can be smarter and check if our table succeeded
+			if (error) {
+				if (completion) {
+					[self.callbackQueue addOperationWithBlock:^{
+						completion(error);
+					}];
+				}
+			} else {
+				// Check again if we have new pending ops while we synced, and repeat as needed
+				[pushQueue_ addOperation:pull];
```
Comment (phvannor) [inline sdk/iOS/src/MSSyncContext.m]: This will be a change in behavior.  We aren't checking the if table is dirty again, just assuming we are good to go on success.  It probably has minimal impact, but the comments on the functions are now wrong.

**Comment by phvannor** [comment]
I expected to see an update to sdk/iOS/src/WindowsAzureMobileServices.h to list the Queue*Operation files, as well as adding the  {ATTRIBUTES = (Public, ) option in the pbxproj file.

**Comment by damienpontifex** [comment]
@shrishrirang that seems like a good solution. I have tested that and seems to work as well.

File: sdk/iOS/src/MSSyncContext.m (hunk 2 above)
Comment (damienpontifex) [inline sdk/iOS/src/MSSyncContext.m]: I couldn't figure out how to recheck if the table was dirty and still utilise the pull operation we returned to the user. 

The only was I cam up with was extracting this dirty/push check into another private method which took the existing pull operation. Thus we could call it again in the callback and pass the contained push operation. Thoughts?

File: sdk/iOS/src/MSSyncContext.m (hunk 1 above)
Comment (damienpontifex) [inline sdk/iOS/src/MSSyncContext.m]: Originally I didn't see a way to return the pull operation from an async queue. Then I moved the pull constructor up to the first expression and I see I can add the async call back in. Will do it.

File: sdk/iOS/src/MSSyncContext.m (hunk 2 above)
Comment (phvannor) [inline sdk/iOS/src/MSSyncContext.m]: I'm OK with the current logic, and only doing 1 push for now, after I wrote this, I ended up adding code to handle pending ops post pull anyways, so its not really necessary to even do the 1 push.  In addtion cancelling the pull NSOperation, should cancel the dependent Push, so this whole section may need reworked later on anyways.  For now, just fix the docs on this function to indicate 1 push will be attempted if pending ops exist for the table.

File: sdk/iOS/src/MSSyncContext.h [hunk 3]
```diff
@@ -8,6 +8,7 @@
 
 @class MSQuery;
 @class MSSyncContext;
+@class MSQueuePushOperation;
```
Comment (shrishrirang) [inline sdk/iOS/src/MSSyncContext.h]: Is this change needed now (in this file and other files)?

File: sdk/iOS/src/MSSyncContext.m [hunk 4]
```diff
@@ -82,7 +82,7 @@ -(NSUInteger) pendingOperationsCount
 /// Begin sending pending operations to the remote tables. Abort the push attempt whenever any single operation
 /// recieves an error due to network or authorization. Otherwise operations will all run and all errors returned
 /// to the caller at once.
--(void) pushWithCompletion:(MSSyncBlock)completion
+-(MSQueuePushOperation *) pushWithCompletion:(MSSyncBlock)completion
```
Comment (shrishrirang) [inline sdk/iOS/src/MSSyncContext.m]: NSOperation* return type?

File: sdk/iOS/src/MSSyncContext.m [hunk 5]
```diff
@@ -404,54 +406,55 @@ - (void) pullWithQuery:(MSQuery *)query queryId:(NSString *)queryId completion:(
...
-                                                                                maxRecords:maxRecords
-                                                                             dispatchQueue:writeOperationQueue
-                                                                             callbackQueue:self.callbackQueue
-                                                                                completion:completion];
-            
-            
-            [pushQueue_ addOperation:pull];
-        }
-    });
+	MSQueuePullOperation *pull = [[MSQueuePullOperation alloc] initWithSyncContext:self
```
Comment (shrishrirang) [inline sdk/iOS/src/MSSyncContext.m]: Minor: Can you fix the formatting?

File: sdk/iOS/src/MSSyncContext.m [hunk 6]
```diff
@@ -404,54 +406,55 @@ - (void) pullWithQuery:(MSQuery *)query queryId:(NSString *)queryId completion:(
...
+				// For now we just abort the pull if the push failed to complete successfully
+				// Long term we can be smarter and check if our table succeeded
+				if (error) {
+					if (completion) {
+						[self.callbackQueue addOperationWithBlock:^{
+							completion(error);
+						}];
+					}
+				} else {
+					// Check again if we have new pending ops while we synced, and repeat as needed
```
Comment (shrishrirang) [inline sdk/iOS/src/MSSyncContext.m]: This comment is not valid anymore as we are not calling pullWithQueryInternal again. Can you update it?

File: sdk/iOS/src/MSSyncContext.m [hunk 7]
```diff
@@ -404,54 +406,55 @@ - (void) pullWithQuery:(MSQuery *)query queryId:(NSString *)queryId completion:(
...
+		// Before we can pull from the remote, we need to make sure out table doesn't having pending operations
+		NSArray *tableOps = [self.operationQueue getOperationsForTable:query.table.name item:nil];
+		if (tableOps.count > 0) {
+			NSOperation *push = [self pushWithCompletion:^(NSError *error) {
+				// For now we just abort the pull if the push failed to complete successfully
+				// Long term we can be smarter and check if our table succeeded
+				if (error) {
+					if (completion) {
+						[self.callbackQueue addOperationWithBlock:^{
+							completion(error);
```
Comment (shrishrirang) [inline sdk/iOS/src/MSSyncContext.m]: If push fails, we won't add the pull operation to the queue. Would that not mean that the pull operation would never finish? Should we not explicitly cancel it?

File: sdk/iOS/test/MSSyncTableTests.m [hunk 8]
```diff
@@ -2184,6 +2258,7 @@ -(void) testPurgePendingOpsOnDifferentTableSuccess
             
             done = YES;
         }];
+		XCTAssertNotNil(purge);
```
Comment (shrishrirang) [inline sdk/iOS/test/MSSyncTableTests.m]: Minor: Can you fix indentation?

File: sdk/iOS/src/MSSyncContextInternal.h [hunk 9]
```diff
@@ -7,6 +7,10 @@
 #import "MSOperationQueue.h"
 #import "MSTable.h"
 
+@class MSQueuePushOperation;
+@class MSQueuePullOperation;
```
Comment (shrishrirang) [inline sdk/iOS/src/MSSyncContextInternal.h]: unnecessary now?

File: sdk/iOS/test/MSSyncTableTests.m (hunk 8 above)
Comment (damienpontifex) [inline sdk/iOS/test/MSSyncTableTests.m]: I had this issue with another PR in the past. I use tab indentation, so lays out correctly in Xcode but GitHub adds the extra indentation when displaying it in the diff (maybe due to the plus symbol next to it)

File: sdk/iOS/src/MSSyncContext.m [hunk 10]
```diff
@@ -404,54 +406,56 @@ - (void) pullWithQuery:(MSQuery *)query queryId:(NSString *)queryId completion:(
...
     dispatch_async(writeOperationQueue, ^{
         // Before we can pull from the remote, we need to make sure out table doesn't having pending operations
         NSArray *tableOps = [self.operationQueue getOperationsForTable:query.table.name item:nil];
         if (tableOps.count > 0) {
-            [self pushWithCompletion:^(NSError *error) {
+            NSOperation *push = [self pushWithCompletion:^(NSError *error) {
                 // For now we just abort the pull if the push failed to complete successfully
                 // Long term we can be smarter and check if our table succeeded
                 if (error) {
+                    [pull cancel];
```
Comment (phvannor) [inline sdk/iOS/src/MSSyncContext.m]: Are you sure calling cancel is enough?  Doesn't that just set the isCancelled flag?  The op itself wouldn't set isFinished until it goes to run.  What flag is the dependent Op observing?

Comment (shrishrirang) [inline sdk/iOS/src/MSSyncContext.m]: Cancelling the NSOperation would mark it finished, so the observing task should be fine.

Comment (damienpontifex) [inline sdk/iOS/src/MSSyncContext.m]: The operation wouldn't have even started at this point, so I don't think we need to do anything else. The completion block will be called if the user has provided one to notify them of the failure. 

I could do some testing though with the returned pull operation as a dependency of another operation. If the execution flows through to the waiting operation when this is cancelled, should all be fine.

Comment (phvannor) [inline sdk/iOS/src/MSSyncContext.m]: Can you link to docs?  Everything I see online, says that the behavior of cancel isn't different if the operation is not running.  So if you want to trigger a dependent task, you are responsible for setting the isFinished flag as well.

Comment (damienpontifex) [inline sdk/iOS/src/MSSyncContext.m]: Running through a unit test, you were right @phvannor that without explicitly setting the finished status, the dependent tasks wouldn't be completed.

Latest commit with changes to access internal method for completion plus unit test for this case added.

Comment (shrishrirang) [inline sdk/iOS/src/MSSyncContext.m]: Thanks for adding UT to test this path.
Btw, should we not simply add the pull NSOperation to the queue immediately after making it dependent on the push NSOperation (instead of waiting for the push to complete successfully)? That way you won't need to call 'completeOperation' explicitly. Just calling cancel ( [push cancel] ) will be enough.

Comment (phvannor) [inline sdk/iOS/src/MSSyncContext.m]: Sadly, then the pull possibly already be running, we'd actually need the push op to expose its error state, so the pull op could look at its dependencies & verify they are were ok before running.  Probably worth looking into, but not sure on effort to do so vs what we have now.

Comment (shrishrirang) [inline sdk/iOS/src/MSSyncContext.m]: Right!
